PLAT-846: Add tree view

mihaelpalfi opened the pull request:
[JIRA ticket](https://reciprocitylabs.atlassian.net/browse/PLAT-846)

#### Description
As per JIRA ticket

@chavlji A couple of questions:
- Should selectable be available on the table level as a parameter also?
- Cant figure out how to strech cell (colspan dosent work)

#### ChangeLOG

[Changelog](https://zen-ui.zengrc.com/?path=/docs/changelog--page)


#### Useful links
[Zen UI readme](https://github.com/reciprocity/zen-ui/blob/main/README.md)

[PR review guidelines](https://github.com/reciprocity/zengrc/blob/develop/doc/dev_environment_and_process/pull_request_reviews.md)

[Git commit guidelines](https://github.com/reciprocity/zengrc/blob/develop/doc/dev_environment_and_process/commit_guidelines.md)

chavlji commented:
@mihaelpalfi 
1. no. Let's keep it per row.
2. I have a solution with cell content position: absolute. Can I push it?

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -1,14 +1,125 @@
+    }
+    return null;
+  }
+  traverseOnClose(): void {
```
Rename to what it actually does - `closeDescendants()`

mihaelpalfi replied:
Done

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -1,14 +1,125 @@
+  traverseOnClose(): void {
+    let next = this.element.nextElementSibling as HTMLZenTableRowElement;
+
+    // Get all rows that have depth greater then the parent
```
Code would be a lot more readable and you wouldn't need this comment if you'd extract gathering descendats into separate function:
`this.descendants().forEach(n => n.expanded = false);`

mihaelpalfi replied:
You are right. Done.

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -1,14 +1,125 @@
+
+  @Watch('expanded')
+  async visibleChanged(expanded?: boolean): Promise<void> {
+    this.children().forEach(child => (child.visible = expanded));
```
I think this doesn't work ok. Try it in some story.

mihaelpalfi replied:
Seems it works. Whats the reason you think it dosent?

chavlji replied:
Because it won't close all decendents but just direct children.

mihaelpalfi replied:
Fixed

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -1,14 +1,125 @@
+  @Element() element: HTMLZenTableRowElement;
+
+  /** Show checkbox */
+  @Prop() readonly selectable = false;
```
I'd rename it to `hasCheckbox`

mihaelpalfi replied:
Hmm dont know but for me this is more suited. And in vue it also uses selectable.

chavlji replied:
Do you mean our vue listing? But it's funny, because in description you're stating just that: "Show checkbox" 😂 
I always prefer prop names that doesn't leave doubt. I connect `selectable` with disabled state of whole row...
Anyway, let's vote and see.

chavlji replied:
`selectable` upvoted. Keep it!

mihaelpalfi replied:
Done

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -1,14 +1,125 @@
+  @Prop() readonly selectable = false;
+
+  /** To add custom content  */
+  @Prop() readonly slotted = false;
```
Why do we need this?

mihaelpalfi replied:
Leftover. Wanted to add this if we wand full width of the cell.

chavlji commented on src/components/zen-table/zen-table.stories.mdx:
```diff
@@ -185,6 +187,123 @@ export const Story3 = () => {
+        <sb-zen-table-cell><sb-zen-text>Status</sb-zen-text></sb-zen-table-cell>
+        <sb-zen-table-cell><sb-zen-text>Some description plus some</sb-zen-text></sb-zen-table-cell>
+      </sb-zen-table-row>
+      <sb-zen-table-row depth="2" selectable>
```
Please use `map` loop for such highly repetative stories. It's hard to read and maintain. Create array and map items from it.

mihaelpalfi replied:
Next pr.

chavlji replied:
Please add link to ticket in this case here.

mihaelpalfi replied:
Added to this ticket PLAT-1340

mplesec reviewed (commented):
The implementation is sound, but I've encountered some bugs? And I think one breaking change to current functionality, but I'm not sure it's caused by this PR.

One of the visual tests incorrectly updates the snapshot, as the new snapshot doesn't match what's in storybook:
![Screenshot 2021-03-03 at 09 52 36](https://user-images.githubusercontent.com/9392804/109780200-0e367f80-7c07-11eb-8478-f3ef57a08091.png)

The titles for this story look out of place:
![Screenshot 2021-03-03 at 09 53 06](https://user-images.githubusercontent.com/9392804/109780307-2efed500-7c07-11eb-8241-36c8b2a8cf64.png)

The sticky header needs a background:

https://user-images.githubusercontent.com/9392804/109780385-42aa3b80-7c07-11eb-9064-4f44cc6c7ad1.mov

mihaelpalfi commented on src/components.d.ts:
```diff
@@ -1361,6 +1389,26 @@ declare namespace LocalJSX {
+         */
+        "selectable"?: false;
+        /**
+          * Visible if no !depth or parent.expanded (read-only)
```
Changed.

mihaelpalfi commented on src/components/zen-table-row/zen-table-row.scss:
```diff
@@ -2,7 +2,71 @@
+  padding-left: #{$checkbox-width + $arrow-width};
+}
+
+@for $i from 0 through 32 {
```
For now its ok. We will probably move this so it will be added dynamically.

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.scss:
```diff
@@ -2,3 +2,15 @@
+  width: 2rem;
+}
+
+.expendable {
```
Done.

mihaelpalfi commented:
> The implementation is sound, but I've encountered some bugs? And I think one breaking change to current functionality, but I'm not sure it's caused by this PR.

1. The vrt screnshot is correct. I removed some rows.
2. The miss aligment issue. It looks ok for me. Only the first row left padding is off and have to still figure it out how to change this without braking thee tree css. Will do this in next pr. 
3. Fixed the sticky issue. Added z index and fixed some css on the header row cell.

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -1,14 +1,138 @@
+  /** Show checkbox */
+  @Prop() readonly selectable = false;
+
+  /** Visible if no depth or parent.expanded */
```
You're not reacting to changes of this attr from apps side, so please mark it as `read-only` in the description.

mihaelpalfi replied:
Done

chavlji commented:
Please don't forget to add link to ticket for `Please use map loop for such highly repetative stories...` here

mihaelpalfi commented:
> Please don't forget to add link to ticket for `Please use map loop for such highly repetative stories...` here

https://reciprocitylabs.atlassian.net/browse/PLAT-1340

chavlji reviewed (approved):
🚢 it!
